dartsim/src/SDFFeatures_TEST.cc

```diff
@@ -95,16 +109,196 @@ World LoadWorld(const std::string &_world)
+{
+  ignition::math::Pose3d pose;
+  ::sdf::Errors errors = _semPose.Resolve(pose);
+  EXPECT_TRUE(errors.empty()) << errors;
+  return pose;
+}
+
+static sdf::JointAxis ResolveJointAxis(const sdf::JointAxis &_unresolvedAxis)
+{
+  ignition::math::Vector3d axisXyz;
+  const sdf::Errors resolveAxisErrors = _unresolvedAxis.ResolveXyz(axisXyz);
+  EXPECT_TRUE (resolveAxisErrors.empty()) << resolveAxisErrors;
+
+  sdf::JointAxis resolvedAxis = _unresolvedAxis;
+
+  const sdf::Errors setXyzErrors = resolvedAxis.SetXyz(axisXyz);
+  EXPECT_TRUE(setXyzErrors.empty()) << setXyzErrors;
+
+  resolvedAxis.SetXyzExpressedIn("");
+  return resolvedAxis;
+}
+
+/////////////////////////////////////////////////
+/// Downstream applications, like ign-gazebo, use this way of world construction
+WorldPtr LoadWorldPiecemeal(const std::string &_world)
+{
+  auto engine = LoadEngine();
+  EXPECT_NE(nullptr, engine);
+  if (nullptr == engine)
+    return nullptr;
+
+  sdf::Root root;
+  const sdf::Errors &errors = root.Load(_world);
+  EXPECT_EQ(0u, errors.size()) << errors;
+
+  EXPECT_EQ(1u, root.WorldCount());
+  const sdf::World *sdfWorld = root.WorldByIndex(0);
+  EXPECT_NE(nullptr, sdfWorld);
+
+  sdf::World newWorld;
+  newWorld.SetName(sdfWorld->Name());
+  newWorld.SetGravity(sdfWorld->Gravity());
+  auto world = engine->ConstructWorld(newWorld);
+  if (nullptr == world)
+    return nullptr;
+
+  std::unordered_map<const sdf::Model *, ModelPtr> modelMap;
+  std::unordered_map<const sdf::Link *, LinkPtr> linkMap;
+
+  auto createModel = [&](const sdf::Model *_model,
+                         const sdf::Model *_parentModel = nullptr) {
+    sdf::Model newSdfModel;
+    newSdfModel.SetName(_model->Name());
+    newSdfModel.SetRawPose(ResolveSdfPose(_model->SemanticPose()));
+    newSdfModel.SetStatic(_model->Static());
+    newSdfModel.SetSelfCollide(_model->SelfCollide());
+
+    ModelPtr newModel;
+    if (nullptr != _parentModel)
+    {
+      auto it = modelMap.find(_parentModel);
+      ASSERT_TRUE(it != modelMap.end());
+      newModel = it->second->ConstructNestedModel(newSdfModel);
+    }
+    else
+    {
+      newModel = world->ConstructModel(newSdfModel);
+    }
+
+    EXPECT_NE(nullptr, newModel);
+    if (nullptr != newModel)
+    {
+      modelMap[_model] = newModel;
+    }
+  };
+
+  for (uint64_t i = 0; i < sdfWorld->ModelCount(); ++i)
+  {
+    const auto *model = sdfWorld->ModelByIndex(i);
+    createModel(model);
+    for (uint64_t nestedInd = 0; nestedInd < model->ModelCount(); ++nestedInd)
+    {
+      createModel(model->ModelByIndex(nestedInd), model);
+    }
+  }
+
+  for (auto [sdfModel, physModel] : modelMap)
+  {
+    for (uint64_t li = 0; li < sdfModel->LinkCount(); ++li)
+    {
+      const auto *link = sdfModel->LinkByIndex(li);
+      sdf::Link newSdfLink;
+      newSdfLink.SetName(link->Name());
+      newSdfLink.SetRawPose(ResolveSdfPose(link->SemanticPose()));
+      newSdfLink.SetInertial(link->Inertial());
```

Comment:
```suggestion
      const auto link = sdfModel->LinkByIndex(li);
      sdf::Link newSdfLink;
      newSdfLink.SetName(link->Name());
      newSdfLink.SetRawPose(ResolveSdfPose(link->SemanticPose()));
      newSdfLink.SetInertial(link->Inertial());
```
Can we add a check to make sure `link` isn't `nullptr`?

Also, I don't think we need the `*` in the return type of the `sdfModel->LinkByIndex(li)` call since we are using `auto` and that method returns a `Link *`.

Reply:
Done in https://github.com/ignitionrobotics/ign-physics/pull/220/commits/53ce2b79c4059185820f7d5ded29221a9cabcb2b